nixos/modules/services/hardware/logiops.nix

```diff
@@ -0,0 +1,164 @@
+                    };
+                  },
+                  {
+                    direction: "Right";
+                    mode: "OnRelease";
+                    action =
+                    {
+                      type = "ToggleSmartshift";
+                    }
+                  },
+                  {
+                    direction: "None"
+                    mode: "NoPress"
+                  }
+                );
+              };
+            },
+            {
+              cid: 0xc4;
+              action =
+              {
+                type: "Keypress";
+                keys: ["KEY_A"];
+              };
+            }
+          );
+        });
+      '';
+      description = mdDoc ''
+        Extra configuration that is appended to the end of the file.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc."logid.cfg".text = ''
+      ${cfg.extraConfig}
+    '';
+
+    systemd.services.logiops = {
+      description = "Logitech Configuration Daemon";
+      documentation = [ "https://github.com/PixlOne/logiops" ];
+
+      wantedBy = [ "graphical.target" ];
+
+      startLimitIntervalSec = 0;
+      after = [ "multi-user.target" ];
+      wants = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.logiops}/bin/logid";
+        Restart = "always";
+        # Ensures that the hid_logitech_hidpp module is loaded before logid starts
+        # https://github.com/PixlOne/logiops/issues/269#issuecomment-1012608838
+        ExecStartPre = "${pkgs.kmod}/bin/modprobe hid_logitech_hidpp || :";
+        User = "root";
+
+        RuntimeDirectory = "logiops";
+
+        CapabilityBoundingSet = [ "CAP_SYS_NICE" ];
+        DeviceAllow = [
+          "/dev/uinput rw"
+          # Globbing does not work here
+          # https://www.freedesktop.org/software/systemd/man/252/systemd.resource-control.html
+          "/dev/hidraw0 rw"
+          "/dev/hidraw1 rw"
+          "/dev/hidraw2 rw"
+          "/dev/hidraw3 rw"
+          "/dev/hidraw4 rw"
+          "/dev/hidraw5 rw"
+          "/dev/hidraw6 rw"
+          "/dev/hidraw7 rw"
+          "/dev/hidraw8 rw"
+          "/dev/hidraw9 rw"
+        ];
+        ProtectClock = true;
+        PrivateNetwork = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        PrivateUsers = true;
+        PrivateMounts = true;
+        PrivateTmp = true;
+        RestrictNamespaces = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectControlGroups = true;
+        MemoryDenyWriteExecute = true;
+        RestrictRealtime = true;
+        LockPersonality = true;
+        ProtectProc = "invisible";
+        SystemCallFilter = [ "nice" "@system-service" "~@privileged" ];
+        RestrictAddressFamilies = [ "AF_NETLINK" ];
+        RestrictSUIDSGID = true;
+        IPAddressDeny = [ "any" ];
```

Comment:
I believe this effect is already achieved by `PrivateNetwork`. AFAIK, it will put the service into it's own network namespace, without any other interfaces than loopback.

Reply:
Is this a reference to just the `IPAddressDeny` line? I agree for that one, but from what I can tell from the `systemd` docs  seems like `PrivateNetwork` doesn't overlap with the other hardening settings (i.e. `SystemCallFilter`, `RestrictAddressFamilies`, `RestrictSUIDSGID`).

Reply:
> Is this a reference to just the IPAddressDeny line?

Yes

> [...] `PrivateNetwork` doesn't overlap with the other hardening settings

That's true, this is a valid config and runs without problems. But setting firewall rules when the service is already isolated from the outside world doesn't really do much.